internal/cmd/envoy/envoy_init.go

```diff
@@ -0,0 +1,160 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package envoy
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	bootstrapv3 "github.com/envoyproxy/go-control-plane/envoy/config/bootstrap/v3"
+	clusterv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	corev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	endpointv3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/durationpb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+const (
+	DefaultEnvoyInitConfigDir      = "/envoyconfigs"
+	DefaultEnvoyInitConfigFilename = "config.json"
+)
+
+func EnvoyInit(configPath string, zoneOverride string) error {
+	if zoneOverride != "" {
+		return writeConfig(configPath, zoneOverride)
+	}
+
+	nodeName := os.Getenv("NODE_NAME")
+
+	if nodeName == "" {
+		return fmt.Errorf("NODE_NAME environment variable is required")
+	}
+
+	node, err := getNode(nodeName)
+	if err != nil {
+		return fmt.Errorf("error getting node %q: %w", nodeName, err)
+	}
+
+	zone, err := buildLocalityZone(node, zoneOverride)
+	if err != nil {
+		return fmt.Errorf("error getting node topology zone: %w", err)
+	}
+
+	// Write locality information to envoy config file
+	if err = writeConfig(configPath, zone); err != nil {
+		return fmt.Errorf("error writing config file: %w", err)
+	}
+
+	return nil
+}
+
+func getNode(nodeName string) (*corev1.Node, error) {
+	c, err := getClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build clientset: %w", err)
+	}
+	return c.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+}
+
+func getClient() (kubernetes.Interface, error) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(config)
+}
+
+// buildLocalityZone configures the envoy locality zone using the Kubernetes node topology labels.
+func buildLocalityZone(node *corev1.Node, override string) (string, error) {
+	if override != "" {
+		return override, nil
+	}
+
+	zone, exists := node.Labels[corev1.LabelTopologyZone]
+	if !exists {
+		return "", fmt.Errorf("zone label %q not found on node %q", corev1.LabelTopologyZone, node.Name)
+	}
+	return zone, nil
+}
+
+// writeConfig writes the locality information to the config file used for Envoy bootstrapping.
+func writeConfig(configPath, zone string) error {
```

Comment:
this is no longer needed, right ?

Reply:
This is still needed so that Envoy knows what the local zone is. The alternative would've been to grab _every_ node and make a fake locality endpoint to represent each zone.

Reply:
I had assumed that `--service-zone` would be enough with the new feature you added in the upstream

Reply:
It makes things easier but it's still necessary unless there's a better idea. --service-zone still needs to have its value set but isn't known until _after_ the point of scheduling. Init container or EDS are the only options I'm aware of.